### PR TITLE
cache templates, refactor inheritability, `optional` documentable flag

### PR DESCRIPTION
**Pull request by anru**
- let documentables be optional
- added `isDocumentableInheritable` method for ability to add new documentables which could be inheritable
- added `getTemplateEngine` method
- cache compiled templates

### CODE REVIEW

**Comment by mourner** [comment]
Hey, nice to see a new contribution from you Andrey!

File: src/leafdoc.js [hunk 1]
```diff
@@ -114,6 +114,13 @@ Leafdoc.prototype.registerDocumentable = function(name, label) {
 	return this;
 };
 
+// 🍂method getTemplateEngine(): Handlebars
+// Returns handlebars template engine used to render templates.
+// You can use it for override helpers or register new.
+Leafdoc.prototype.getTemplateEngine = function() {
```
Comment (IvanSanchez) [inline src/leafdoc.js:133]: What's a use case for this? It's not clear for me what this feature can be used for.

File: src/leafdoc.js [hunk 2]
```diff
@@ -395,17 +402,18 @@ Leafdoc.prototype.addStr = function(str, isSource) {
...
 					if (content) {
 						var split = regexps.functionDefinition.exec(content);
 						if (!split) {
 							console.error('Invalid ' + directive + ' definition: ', content);
 						} else {
 							name = split[1];
-							paramString = split[2];
-							type = split[3];
-							defaultValue = split[4];
+							optional = split[2] == '?';
```
Comment (IvanSanchez) [inline src/leafdoc.js:426]: Perhaps you can provide a template which uses the `optional` property? That'd illustrate it better.

File: src/leafdoc.js [hunk 3]
```diff
@@ -567,7 +576,17 @@ Leafdoc.prototype._stringifyNamespace = function(namespace, isMini) {
 	});
 };
 
-
+/*
+ * 🍂method isDocumentableInheritable(name: String): Boolean
+ * Determines documentable can be inherited.
+ */
+Leafdoc.prototype.isDocumentableInheritable = function(name) {
```
Comment (IvanSanchez) [inline src/leafdoc.js]: What about custom documentables via `registerDocumentable()`? It'd be nice to specify if custom documentables are inheritable or not. Maybe have the inheritable ones in an array.

**Comment by IvanSanchez** [comment]
Looks like a nice set of new features! Thanks! :-)

However, I don't feel the changes are 100% ready - see a few comments I've left about some of the code.

File: src/leafdoc.js (hunk 1 above)
Comment (anru) [inline src/leafdoc.js:133]: My use case is override default `type` helper to better one in order to support complex types, for example `Array<L.LatLng>`. Like this:

``` js
function makeTypesBetter(handlebars) {
  const makeLink = handlebars.helpers.type;
  const typeRe = /[\w\._\s]+/g;

  handlebars.registerHelper('type', function(str) {
    return str
      .replace(/[\<\>]/g, function(str) {
        return escapeHtml(str);
      })
      .replace(typeRe, function(typeStr) {
        return makeLink(typeStr);
      });
  });
}

makeTypesBetter(doc.getTemplateEngine());
```

Btw, we can add support for complex types in a primary codebase, but nonetheless having ability to extend built-in helpers and add new is good idea I think.

Comment (IvanSanchez) [inline src/leafdoc.js:133]: Aaaah, I got it. That's a nifty idea!

I think we'll need to add some proper documentation about how the templates work, but that can be done later.

File: src/leafdoc.js (hunk 2 above)
Comment (anru) [inline src/leafdoc.js:426]: Sure. Suppose we have some data structures for our api. And I want provide default value and optionality:

``` handlebars
<table><thead>
    <tr>
        <th>Field</th>
        <th>Type</th>
        <th>Required</th>
        <th>Example</th>
        <th>Description</th>
    </tr>
    </thead><tbody>
    {{#each documentables}}
    <tr id='{{id}}'>
        <td><code><b>{{name}}</b></code></td>
        <td><code>{{{type type}}}</code>
        <td>{{#unless optional}}Yes{{/unless}}</td>
        <td><code>{{defaultValue}}</code></td>
        <td>{{{markdown comments}}}</td>
    </tr>
    {{/each}}
</tbody></table>
```

Comment (IvanSanchez) [inline src/leafdoc.js:426]: Maybe we can extend the default templates with this?

File: src/leafdoc.js (hunk 3 above)
Comment (anru) [inline src/leafdoc.js]: For myself I use following code:

``` es6
class ApiLeafdoc extends LeafDoc {
  isDocumentableInheritable(name) {
    return super.isDocumentableInheritable(name) || name === 'field';
  }
}
```

I don't mind having inheritable documentables in array, in this case we can simplify code for add new inheritable documentables:

``` js
doc.inheritableDocumentables.push('field');
```

What do you think ? Shall I use this option?

Comment (IvanSanchez) [inline src/leafdoc.js]: No, I'd rather have something like `doc.registerDocumentable(name, label, inheritable)`.

File: src/leafdoc.js (hunk 2 above)
Comment (anru) [inline src/leafdoc.js:426]: Ok, will do

File: src/leafdoc.js (hunk 3 above)
Comment (anru) [inline src/leafdoc.js]: :+1:

File: src/leafdoc.js (hunk 2 above)
Comment (anru) [inline src/leafdoc.js:426]: Hm, not sure how it should looks like. Optional flag makes sense only for fields in data structures for example. But not for options, they are optional already.

Comment (anru) [inline src/leafdoc.js:426]: cc @IvanSanchez

**Comment by anru** [comment]
Heey, are here somebody? :)

**Comment by IvanSanchez** [comment]
Oh, I forgot about this :-D

**Comment by IvanSanchez** [comment]
Looks good. I'll bump the version, do the `npm publish` magic, and see how it goes :-)

**Pull request merged**